rasa/cdu/command_processor.py

```diff
@@ -0,0 +1,249 @@
+from typing import List, Set, Type
+
+import structlog
+from rasa.cdu.commands import (
+    CancelFlowCommand,
+    Command,
+    CorrectSlotCommand,
+    HandleInterruptionCommand,
+    ListenCommand,
+    SetSlotCommand,
+    StartFlowCommand,
+    command_from_json,
+)
+from rasa.cdu.flow_stack import FlowStack, FlowStackFrame, StackFrameType
+from rasa.shared.core.constants import (
+    CANCELLED_FLOW_SLOT,
+    CORRECTED_SLOTS_SLOT,
+    FLOW_STACK_SLOT,
+)
+from rasa.shared.core.events import Event, SlotSet, UserUttered
+from rasa.shared.core.flows.flow import FlowsList, QuestionFlowStep
+from rasa.shared.core.trackers import DialogueStateTracker
+from rasa.shared.nlu.constants import COMMANDS
+
+
+structlogger = structlog.get_logger()
+
+FLOW_PATTERN_CORRECTION_ID = "pattern_correction"
+
+FLOW_PATTERN_CANCEl_ID = "pattern_cancel_flow"
+
+FLOW_PATTERN_LISTEN_ID = "pattern_listen"
+
+
+def contains_command(commands: List[Command], typ: Type[Command]) -> bool:
+    """Check if a list of commands contains a command of a given type."""
+    return any(isinstance(command, typ) for command in commands)
+
+
+def _slot_sets_after_latest_message(tracker: DialogueStateTracker) -> List[SlotSet]:
+    """Get all slot sets after the latest message.
+
+    Args:
+        tracker: The tracker to get the slot sets from.
+
+    Returns:
+    All slot sets after the latest user message.
+    """
+    if not tracker.latest_message:
+        return []
+
+    slot_sets = []
+
+    for event in reversed(tracker.applied_events()):
+        if isinstance(event, UserUttered):
+            break
+        elif isinstance(event, SlotSet):
+            slot_sets.append(event)
+    return slot_sets
+
+
+def _get_commands_from_tracker(tracker: DialogueStateTracker) -> List[Command]:
+    """Extracts the commands from the tracker.
+
+    Args:
+        tracker: The tracker containing the conversation history up to now.
+
+    Returns:
+        The commands.
+    """
+    if tracker.latest_message:
+        dumped_commands = tracker.latest_message.parse_data.get(COMMANDS) or []
+        assert isinstance(dumped_commands, list)
+        return [command_from_json(command) for command in dumped_commands]
+    else:
+        return []
+
+
+def execute_commands(
+    tracker: DialogueStateTracker, all_flows: FlowsList
+) -> List[Event]:
+    """Executes a list of commands.
+
+    Args:
+        commands: The commands to execute.
+        tracker: The tracker to execute the commands on.
+        all_flows: All flows.
+
+    Returns:
+    A tuple of the action to execute and the events that were created.
+    """
+    commands: List[Command] = _get_commands_from_tracker(tracker)
+    flow_stack = FlowStack.from_tracker(tracker)
+    original_stack_dump = flow_stack.as_dict()
+
+    current_top_flow = flow_stack.top_flow(all_flows)
+
+    commands = clean_up_commands(commands, tracker, all_flows)
+
+    events: List[Event] = []
+
+    # TODO: should this really be reversed? 🤔
```

Comment:
@twerkmeister wdyt?

Reply:
so cancel does not cancel the top most flow at the time it gets excuted but the top most flow at the time of the llm prediction. so we already check that we cancel the right flow. that should work

Reply:
mhm I could fix the start flow & chitchat one in "clean up commands". but we don't have a chitchat atm do we 😂

Reply:
I think for now this works, but something to keep in mind as this could create issues in the future.